mllib/src/main/scala/org/apache/spark/ml/tree/treeModels.scala

```diff
@@ -78,6 +78,28 @@ private[spark] trait DecisionTreeModel {
 
   /** Convert to spark.mllib DecisionTreeModel (losing some information) */
   private[spark] def toOld: OldDecisionTreeModel
+
+  /** Returns an iterator that traverses (DFS, left to right) the leaves
+   *  in the subtree of this node.
+   */
+  private def leafIterator(node: Node): Iterator[LeafNode] = {
+    node match {
+      case l: LeafNode => Iterator.single(l)
+      case n: InternalNode =>
+        leafIterator(n.leftChild) ++ leafIterator(n.rightChild)
+    }
+  }
+
+  @transient private lazy val leafIndices: Map[LeafNode, Int] = {
```

Comment:
I am a bit worried by the extra memory pressure caused by this, in case trees are big.... can't we rather assign an index to the `LeafNode` class?

Reply:
I had impled another leaf-transformation in the .mllib side https://github.com/apache/spark/pull/11520/files, and it used the sorted `leafId` as the output.
However, in the .ml side, the `LeafNode` class do not contain a Id, and is exposed to the end user. So I tend to leave current `LeafNode` class alone.
As to the extra memory pressure, I think it is much smaller than the model itself.
WDYT @srowen

Reply:
If it's lazy, it won't affect usages that don't set the leaf col (right?). I think it's reasonable.

Reply:
yes, this var is only generated when needed. and I make it transient, so it will cause extra network/storage overhead.